### PR TITLE
Update CONTRIBUTION.md

### PR DESCRIPTION
**Pull request by MichaelHillcox**
Check it over see if you agree. Only small changes

### CODE REVIEW

File: CONTRIBUTION.md [hunk 1]
```diff
@@ -1,42 +1,52 @@
...
     * `public void emptyImpl() {}`
+    * Also try and avoid introducing code that may require empty methods
 * Method/Class annotations should gain their own lines in front of the definition
 * Parameter Annotations should be surrounded by exactly one space
+* Comments shoud always start with a single space `// comment`
+
 ### Code Style
 In general apply common sense to the Code you write. "Effective Java" is also always a good start.
 * Avoid unnecessary else clauses
+* Avoid large case statements and it's likely not needed. If you find yourself requiring a large case statement block then try to think about how you might be able to refactor your code in a way that avoids the requirement. Small case statements are fine.
```
Comment (MajorTuvok) [inline CONTRIBUTION.md]:
```suggestion
* Avoid large case statements as it's likely not needed. If you find yourself requiring a large case statement block then try to think about how you might be able to refactor your code in a way that avoids the requirement. Small case statements are fine.
```

File: CONTRIBUTION.md [hunk 2]
```diff
@@ -52,6 +62,8 @@ In general apply common sense to the Code you write. "Effective Java" is also al
 * Use Guava's `Preconditions` class for Argument validation
 * Validate `@NonNull` values with `Objects.requireNonNull` if necessary
 * use getters and setters, even in the declaring class
+* avoid getters and setters for protected methods
```
Comment (MajorTuvok) [inline CONTRIBUTION.md]:
```suggestion
* avoid getters and setters for protected fields
```
Also this needs clarification that in the API, the reverse holds true...

File: CONTRIBUTION.md [hunk 3]
```diff
@@ -52,6 +62,8 @@ In general apply common sense to the Code you write. "Effective Java" is also al
...
+* use protected methods where external classes do not require access or in instances where you have a private var without getters or setters.
```
Comment (MajorTuvok) [inline CONTRIBUTION.md]: I'm unsure what you're referring to here actually - In General access should be as limited as possible...? 

package-private Methods or classes are also feasible things...

File: CONTRIBUTION.md (hunk 2 above)
Comment (MichaelHillcox) [reply CONTRIBUTION.md]: Will review when we get an API?

File: CONTRIBUTION.md (hunk 3 above)
Comment (MichaelHillcox) [reply CONTRIBUTION.md]: Effectively use the right level of access

File: CONTRIBUTION.md (hunk 2 above)
Comment (MajorTuvok) [reply CONTRIBUTION.md]: I'd rather have it in here already, as we have the beginning in 1.13 already and hn's PR target's API code so... There are already PR's targeting the API...

File: CONTRIBUTION.md (hunk 3 above)
Comment (MajorTuvok) [reply CONTRIBUTION.md]:
```suggestion
* In general try to use the smallest visibility-scope possible, expcept if it's likely that others will want access to a given method.
```

File: CONTRIBUTION.md (hunk 2 above)
Comment (MajorTuvok) [reply CONTRIBUTION.md]:
```suggestion
* Avoid getters and setters for protected fields in non API-Code. 
* Avoid non-private variables in API-Code and provide getters/setters instead. Bear in mind that this does not apply to constants (`public static final type xyz = ...`) though.
```

**Comment by MichaelHillcox** [comment]
I'll get @Phylogeny to do a read over then I'll squish and merge :+1:

**Comment by hnOsmium0001** [comment]
I think you can apply suggestions with the Web Interface in batches?

**Comment by MichaelHillcox** [comment]
I'll be squashing them, not an issue :P

File: CONTRIBUTION.md [hunk 4]
```diff
@@ -1,42 +1,52 @@
...
 We know that as of 1.12 this is not the case for great parts of the mod, but we are trying to improve upon the code-base.
-Also your PR should not contain any formatting only changes to Files, as this simply makes reviewing them unnecessarily complex.
-Try to avoid unnecessary commits in order not to pollute the commit history.
+Unless your Pr is directly related to code cleaning then your PR should not contain any formatting changes to unchanged code blocks as this simply makes reviewing them unnecessarily complex.
+
 ### Formatting
 * Use 4 space indent instead of Tabs
 * operators should be surrounded by spaces
-* a `,` comma to separate types should be followed by exactly one Space
+  * For example `if(statement == true)`
```
Comment (hnOsmium0001) [inline CONTRIBUTION.md]: Shouldn't there be a space in front of the `if` statement? Or we have not yet agreed on that one?

File: CONTRIBUTION.md [hunk 5]
```diff
@@ -1,42 +1,52 @@
...
 * A reference to the corresponding issue or a full description of the feature being implemented
 * An explanation how this PR attempts to implement the given feature
 * If testing was performed, what tests and how extensive they have been
+
 ## Code Guidelines
 In General your code should be well-structured and easy to read. 
 We know that as of 1.12 this is not the case for great parts of the mod, but we are trying to improve upon the code-base.
-Also your PR should not contain any formatting only changes to Files, as this simply makes reviewing them unnecessarily complex.
-Try to avoid unnecessary commits in order not to pollute the commit history.
+Unless your Pr is directly related to code cleaning then your PR should not contain any formatting changes to unchanged code blocks as this simply makes reviewing them unnecessarily complex.
```
Comment (hnOsmium0001) [inline CONTRIBUTION.md]: Consider using capitalized `PR`?

Comment (hnOsmium0001) [inline CONTRIBUTION.md]: Consider using a capitalized `PR`?
```suggestion
Unless your PR is directly related to code cleaning then your PR should not contain any formatting changes to unchanged code blocks as this simply makes reviewing them unnecessarily complex.
```

File: CONTRIBUTION.md [hunk 6]
```diff
@@ -1,42 +1,52 @@
...
 * `{` should not be prefixed with returns
-* Methods should have exactly one free line between them (/their docs)
+* Methods should have exactly one free line between them (or their doc blocks)
 * Avoid empty lines at the start of class, interface or enum definitions
 * Avoid unnecessary lines in empty Methods
     * `public void emptyImpl() {}`
+    * Also try and avoid introducing code that may require empty methods
 * Method/Class annotations should gain their own lines in front of the definition
 * Parameter Annotations should be surrounded by exactly one space
+* Comments shoud always start with a single space `// comment`
```
Comment (hnOsmium0001) [inline CONTRIBUTION.md]: Capitalize as said in line 55 (Code Style)
> * Capitalize comments

and fix typo "shoud".

```suggestion
* Comments should always start with a single space `// Comment`
```

File: CONTRIBUTION.md (hunk 4 above)
Comment (Phylogeny) [reply CONTRIBUTION.md]: Yes, a space after the `if` is definitely the established convention in the code base.

File: CONTRIBUTION.md [hunk 7]
```diff
@@ -52,6 +62,9 @@ In general apply common sense to the Code you write. "Effective Java" is also al
 * Use Guava's `Preconditions` class for Argument validation
 * Validate `@NonNull` values with `Objects.requireNonNull` if necessary
 * use getters and setters, even in the declaring class
+* Avoid getters and setters for protected fields in non API-Code. 
+* Avoid non-private variables in API-Code and provide getters/setters instead. Bear in mind that this does not apply to constants (`public static final type xyz = ...`) though.
```
Comment (hnOsmium0001) [inline CONTRIBUTION.md]: I think type should be capitalized like `Type` or just use `T`.
```suggestion
* Avoid non-private variables in API-Code and provide getters/setters instead. Bear in mind that this does not apply to constants (`public static final Type xyz = ...`) though.
```

File: CONTRIBUTION.md [hunk 8]
```diff
@@ -1,42 +1,52 @@
...
     * `public void emptyImpl() {}`
+    * Also try and avoid introducing code that may require empty methods
 * Method/Class annotations should gain their own lines in front of the definition
 * Parameter Annotations should be surrounded by exactly one space
+* Comments shoud always start with a single space `// comment`
+
 ### Code Style
 In general apply common sense to the Code you write. "Effective Java" is also always a good start.
 * Avoid unnecessary else clauses
+* Avoid large case statements as it's likely not needed. If you find yourself requiring a large case statement block then try to think about how you might be able to refactor your code in a way that avoids the requirement. Small case statements are fine.
```
Comment (hnOsmium0001) [inline CONTRIBUTION.md]:
```suggestion
* Avoid large case statements as it's likely not needed. If you find yourself requiring a large case statement block then try to think about how you might be able to refactor your code in a way that avoids the requirement. Small case statements are fine.
    * In most cases this can be avoided by giving the object that's being tested on an abstract method to be overridden. For example, give item and block objects a method called `getGuiElement(EntityPlayer, int, int ,int)` and call it in `GuiProxy` instead of using a case statement.
    * The other approach would be using a `Map<MyObject,Comsumer<MyObject>>` and setup all different cases as lambdas where the parameter is the object being tested.
```
Just some personal opinion.

File: CONTRIBUTION.md [hunk 9]
```diff
@@ -52,6 +62,9 @@ In general apply common sense to the Code you write. "Effective Java" is also al
 * Use Guava's `Preconditions` class for Argument validation
 * Validate `@NonNull` values with `Objects.requireNonNull` if necessary
 * use getters and setters, even in the declaring class
+* Avoid getters and setters for protected fields in non API-Code.
```
Comment (Phylogeny) [inline CONTRIBUTION.md]:
```suggestion
* Use protected fields in non-API superclasses, rather than private ones with getters and setters, if
      1. Those field are barely accessed in that superclass and are heavily accessed in its subclasses.
      2. There is no additional need for getters and setters, such as:
          1.  External access to the field .
          2. The need to run some additional code or logic when getting or setting the value of the field.
```
The point I was making was much more narrow that that. The crux of my argument was that without some additional need for getters/setters, if a field is hardly used the superclass and heavily used in its subclasses:
1.  Is effectively the property of the subclass, thus it makes sense not to have the superclass horde it away under complete privatized protection.
2. To privatize it means either:
    1. Declaring local variables in subclass methods, which just needlessly takes up space and requires needless overhead.
    2. Peppering the subclass methods with getters that accomplish nothing but add the word `get` and the symbols `()` everywhere. `getX()` is longer than `x`, and objectively just means adding unnecessary length to the lines of code that access the field.

The **_sole_** accomplishment of privatizing a protected field, in the case I described above, is to:
1. Add 2 methods (the getter & setter)
2. Add `get` and `()` everywhere.

The bottom line is that doing so simply adds unnecessary text without logical justification.

File: CONTRIBUTION.md (hunk 4 above)
Comment (MichaelHillcox) [reply CONTRIBUTION.md]: I'm happy to include this as it is inline with these guidelines but I think if we're going to be aiming for code readbility then we should also add a space before the ) and after the ( 

So from this 
`if (statement == true)`
to 
`if ( statement == true )`

Comment (MajorTuvok) [reply CONTRIBUTION.md]: If you ask me this has the opposite effect... @Phylogeny ?

**Comment by MichaelHillcox** [comment]
I'll merge after this comment has been addressed https://github.com/Direwolf20-MC/BuildingGadgets/pull/267#discussion_r265928431

File: CONTRIBUTION.md (hunk 1 above)
Comment (Phylogeny) [reply CONTRIBUTION.md]: I'm also of the preference (aesthetic and otherwise) for no spaces.

File: CONTRIBUTION.md (hunk 4 above)
Comment (Phylogeny) [reply CONTRIBUTION.md]: I'm also of the preference (aesthetic and otherwise) for no spaces.

Comment (MichaelHillcox) [reply CONTRIBUTION.md]: Eww, well, majority rules in favour of not me :P
